Frameworks/Foundation/NSInvocation.mm

```diff
@@ -14,71 +14,93 @@
 //
 //******************************************************************************
 
-#import "Starboard.h"
-#import "Foundation/NSInvocation.h"
+#import <Starboard.h>
+#import <Foundation/NSInvocation.h>
 #import <ctype.h>
-#import "LoggingNative.h"
 #import <objc/encoding.h>
+#import "LoggingNative.h"
 #import "ErrorHandling.h"
 
+#import "NSMethodSignatureInternal.h"
+#import "NSInvocationInternal.h"
+
+#include <vector>
+#include <memory>
+
 static const wchar_t* TAG = L"NSInvocation";
 
-@implementation NSInvocation
-static void* copyArgument(NSInvocation* self, void* buf, int index) {
-    char* type = (char*)[self->_methodSignature getArgumentTypeAtIndex:index];
+@implementation NSInvocation {
+    NSMethodSignature* _methodSignature;
 
-    int size = objc_sizeof_type(type);
+    std::unique_ptr<_NSInvocationCallFrame> _callFrame;
 
-    void* ret = (void*)IwMalloc(size);
-    memcpy(ret, buf, size);
+    size_t _returnLength;
+    BOOL _retainArguments;
 
-    return ret;
+    uint8_t _smallReturnValueOptimization[16];
+    void* _returnValue;
 }
 
 /**
  @Status Interoperable
 */
-+ (NSInvocation*)invocationWithMethodSignature:(id)methodSignature {
-    NSInvocation* ret = [self alloc];
-    ret->_methodSignature = [methodSignature retain];
++ (NSInvocation*)invocationWithMethodSignature:(NSMethodSignature*)methodSignature {
+    return [[[self alloc] initWithMethodSignature:methodSignature] autorelease];
+}
+
+/**
+ @Status Interoperable
+ */
+- (instancetype)initWithMethodSignature:(NSMethodSignature*)methodSignature {
+    if (self = [super init]) {
+        _methodSignature = [methodSignature retain];
+
+        _callFrame.reset(new _NSInvocationCallFrame(methodSignature));
 
-    return [ret autorelease];
+        // true _non-promoted_ return length.
+        // Used only for copying the argument back to the caller.
+        _returnLength = methodSignature.methodReturnLength;
+
+        auto promotedReturnLength = _callFrame->getReturnLength();
+        if (promotedReturnLength > _countof(_smallReturnValueOptimization)) {
```

Comment:
>_countof [](start = 35, length = 8)

nit: sizeof may be more flexible here in case _smallReturnValueOptimization ever changes to anything other than a byte array.  _countof usage here threw me off a bit.